include/swift/AST/ConstTypeInfo.h

```diff
@@ -199,6 +200,37 @@ class TypeValue : public CompileTimeValue {
   swift::Type Type;
 };
 
+/// A representation of a Keypath
+class KeyPathValue : public CompileTimeValue {
+public:
+  struct Component {
+    std::string Label;
+    swift::Type Type;
+  };
+  KeyPathValue(std::string Path,
+               swift::Type RootType,
+               swift::Type ValueType,
+               std::vector<Component> Components)
+  : CompileTimeValue(ValueKind::KeyPath), Path(Path), RootType(RootType), ValueType(ValueType), Components(Components) {}
+
+  std::string getPath() const { return Path; }
+  swift::Type getRootType() const { return RootType; }
+  swift::Type getValueType() const { return ValueType; }
+  std::vector<Component> getComponents() const {
+    return Components;
+  }
+
+  static bool classof(const CompileTimeValue *T) {
+    return T->getKind() == ValueKind::KeyPath;
+  }
+
+private:
+  std::string Path;
+  swift::Type RootType;
+  swift::Type ValueType;
```

Comment:
Is this different for a KeyPath than the type of the last component?

Reply:
Hmm, I cannot think of a case for which that would be true, but I still added this here just in case there are cases that I am not thinking of.